gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/MessageConsumerImpl.java

```diff
@@ -0,0 +1,281 @@
+   */
+  static class DefaultExecutorFactory implements ExecutorFactory<ExecutorService> {
+
+    private final ExecutorService executor = Executors.newSingleThreadExecutor();
+
+    @Override
+    public ExecutorService get() {
+      return executor;
+    }
+
+    @Override
+    public void release(ExecutorService executor) {
+      executor.shutdownNow();
+    }
+  }
+
+  private MessageConsumerImpl(Builder builder) {
+    this.pubsubOptions = builder.pubsubOptions;
+    this.subscription = builder.subscription;
+    this.messageProcessor = builder.messageProcessor;
+    this.pubsubRpc = pubsubOptions.rpc();
+    this.pubsub = pubsubOptions.service();
+    this.deadlineRenewer = builder.deadlineRenewer;
+    this.queuedCallbacks = new AtomicInteger();
+    this.timer = SharedResourceHolder.get(TIMER);
+    this.executorFactory = firstNonNull(builder.executorFactory, new DefaultExecutorFactory());
+    this.executor = executorFactory.get();
+    this.maxQueuedCallbacks = firstNonNull(builder.maxQueuedCallbacks, MAX_QUEUED_CALLBACKS);
+    this.scheduleRunnable = new Runnable() {
+      @Override
+      public void run() {
+        synchronized (futureLock) {
+          if (closed) {
+            return;
+          }
+          pull();
+        }
+      }
+    };
+    nextPull();
+  }
+
+  private Runnable ackingRunnable(final ReceivedMessage receivedMessage) {
+    return new Runnable() {
+      @Override
+      public void run() {
+        try {
+          messageProcessor.process(receivedMessage);
+          pubsub.ackAsync(receivedMessage.subscription(), receivedMessage.ackId());
+        } catch (Exception ex) {
+          pubsub.nackAsync(receivedMessage.subscription(), receivedMessage.ackId());
+        } finally {
+          deadlineRenewer.remove(receivedMessage.subscription(), receivedMessage.ackId());
+          queuedCallbacks.decrementAndGet();
+          // We can now pull more messages. We do not pull immediately to possibly wait for other
+          // callbacks to end
+          scheduleNextPull(500, TimeUnit.MILLISECONDS);
+        }
+      }
+    };
+  }
+
+  private PullRequest createPullRequest() {
+    return PullRequest.newBuilder()
+        .setSubscription(formatSubscriptionName(pubsubOptions.projectId(), subscription))
+        .setMaxMessages(maxQueuedCallbacks - queuedCallbacks.get())
+        .setReturnImmediately(false)
+        .build();
+  }
+
+  private void scheduleNextPull(long delay, TimeUnit timeUnit) {
+    synchronized (futureLock) {
+      if (!closed && stopped) {
+        scheduledFuture = timer.schedule(scheduleRunnable, delay, timeUnit);
+      }
+    }
+  }
+
+  private void nextPull() {
+    synchronized (futureLock) {
+      if (closed) {
+        return;
+      }
+      if (queuedCallbacks.get() == maxQueuedCallbacks) {
+        stopped = true;
+      } else {
+        stopped = false;
+        scheduledFuture = timer.submit(scheduleRunnable);
+      }
+    }
+  }
+
+  private void pull() {
```

Comment:
Do you think code might be simpler if scheduleRunnable's Runnable was an inner class
with this code embedded in the run method (including the synchronized part)?